flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/bundle/InputSideHasNoUniqueKeyBundle.java

```diff
@@ -96,15 +96,19 @@ private boolean foldRecord(RowData joinKey, int hashKey, RowData record) {
             RowData rec = iterator.previous();
             if ((RowDataUtil.isAccumulateMsg(record) && RowDataUtil.isRetractMsg(rec))
                     || (RowDataUtil.isRetractMsg(record) && RowDataUtil.isAccumulateMsg(rec))) {
-                iterator.remove();
-                actualSize--;
-                if (list.isEmpty()) {
-                    bundle.get(joinKey).remove(hashKey);
-                    if (bundle.get(joinKey).isEmpty()) {
-                        bundle.remove(joinKey);
+                // here it's necessary to additionally check that record == rec because hashKey of
+                // these two records might collide
+                if (record.equals(rec)) {
```

Comment:
This is indeed a solution, but I think the one you mentioned in [issue](https://issues.apache.org/jira/browse/FLINK-35184?page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel&focusedCommentId=17839540#comment-17839540) is better.  Current implementation occupies more space compared to the solution mentioned above. WYDT?

Reply:
Actually no. Because my first solution requires to copy the record inside the `getRecords()` method. This is necessary because if we are storing the record as key and `List<RowKind>` as value then we should have `list.size()` of records while sending them to processing. Thats's why we should do smth like `rowSerializer.copy(record)` for every RowKind in list or use absolutely different structure for storing the records and rowKinds. I realized it during implementation.

Reply:
Ok. I agree with you.

Reply:
BTW, the record in list has its own RowKind which needs to be ignored here.

Reply:
BTW, the record in list has its own RowKind which need to be ignored here.

Reply:
> BTW, the record in list has its own RowKind which need to be ignored here.

Nice, thank you for this notice! Corrected by additional changing rowKind to +I and back.